doc/adr/0001-jetstream-json-api-design.md

```diff
@@ -0,0 +1,134 @@
+
+In client responses we will use URL Encoding to provide additional context so multiple fields worth of
+context can be provided. URL Encoding has [broad language support](https://www.rosettacode.org/wiki/URL_decoding).
+
+### Admin APIs
+
+#### Requests
+
+All Admin APIs that today accept `nil` body should also accept an empty JSON document as request body.
+
+#### Responses
+
+All responses will be JSON objects, a few examples will describe it best. Any error that happens has to be 
+communicated within the originally expected message type. Even the case where JetStream is not enabled for
+an account, the response has to be the data type the API would normally generate with additional `error`.
+
+Successful Stream Info:
+
+```json
+{
+  "type": "io.nats.jetstream.api.v1.stream_info",
+  "time": "2020-04-23T16:51:18.516363Z",
+  "config": {
+    "name": "STREAM",
+    "subjects": [
+      "js.in"
+    ],
+    "retention": "limits",
+    "max_consumers": -1,
+    "max_msgs": -1,
+    "max_bytes": -1,
+    "max_age": 31536000,
+    "max_msg_size": -1,
+    "storage": "file",
+    "num_replicas": 1
+  },
+  "state": {
+    "messages": 95563,
+    "bytes": 40104315,
+    "first_seq": 34,
+    "last_seq": 95596,
+    "consumer_count": 1
+  }
+}
+```
+
+Stream Info Error:
+
+```json
+{
+  "type": "io.nats.jetstream.api.v1.stream_info",
+  "time": "2020-04-23T16:51:18.516363Z",
+  "error": {
+    "description": "unknown stream bob",
+    "code": 400
+  },
+  "config": {
+    "name": "",
+    "retention": "",
+    "max_consumers": 0,
+    "max_msgs": 0,
+    "max_bytes": 0,
+    "max_age": 0,
+    "storage": "",
+    "num_replicas": 0
+  },
+  "state": {
+    "messages": 0,
+    "bytes": 0,
+    "first_seq": 0,
+    "last_seq": 0,
+    "consumer_count": 0
+  }
+}
+```
+
+Here we have a minimally correct response with the additional error object.
+
+In the `error` struct we have `description` as a short human friendly explanation that should include enough context to
+identify what Stream or Consumer is being interacted with and whatever else we feel will help the user while not sharing
+privileged account information.  `code` will be HTTP like, 400 human error, 500 server error etc.
+
+Today the list API's just return `["ORDERS"]`, these will become:
+
+```json
+{
+  "type": "io.nats.jetstream.api.v1.stream_list",
+  "time": "2020-04-23T16:51:18.516363Z",
+  "streams": [
+    "ORDERS"
+  ]
+}
+```
+
+With the same `error` treatment when some error happens.
+
+### Client Interactions
+
+To keep clients on small devices viable we will standardise our responses as in the examples below. 
+
+ * `+OK` everything is fine no additional context 
+ * `+OK stream=ORDERS&sequence=10` everything is fine, we have additional data to pass using standard URL encoding
+ * `-ERR` something failed, we have no reason
+ * `-ERR reason=reason%20for%20failure&stream=STREAM` something failed, we have additional context to provide using standard URL encoding
```

Comment:
Maybe within single quotes as the server already does for errors to not change the parsers there (e.g. `-ERR 'Unknown Protocol Operation'`, `-ERR ''`).  Also a NATS client technically can already send custom errors to the server and the server will process them (and log them), but the server will then disconnect the client:

```console
$ telnet 127.0.0.1 4222
-ERR 'foo bar'
[30801] 2020/04/30 09:58:52.720484 [ERR] 127.0.0.1:63165 - cid:4 - Client Error 'foo bar'
```

Reply:
if you wanted to encode more than just a string, you cannot with more difficult parsing.

Reply:
we can surround them with quotes or not, either works for me - as long as we always do whatever one we pick, so since we have quotes in some places shall we say we should add them here too always?